ps3.py

```diff
@@ -0,0 +1,346 @@
+    """
+    Displays the letters currently in the hand.
+
+    For example:
+       display_hand({'a':1, 'x':2, 'l':3, 'e':1})
+    Should print out something like:
+       a x x l l l e
+    The order of the letters is unimportant.
+
+    hand: dictionary (string -> int)
+    """
+
+    for letter in hand.keys():
+        for j in range(hand[letter]):
+            print(letter, end=' ')  # print all on the same line
+    print()  # print an empty line
+
+
+def deal_hand(n):
+    """
+    Returns a random hand containing n lowercase letters.
+    ceil(n/3) letters in the hand should be VOWELS (note,
+    ceil(n/3) means the smallest integer not less than n/3).
+
+    Hands are represented as dictionaries. The keys are
+    letters and the values are the number of times the
+    particular letter is repeated in that hand.
+
+    n: int >= 0
+    returns: dictionary (string -> int)
+    """
+
+    hand = {}
+    num_vowels = int(math.ceil(n / 3))
+
+    for i in range(num_vowels):
+        x = random.choice(VOWELS)
+        hand[x] = hand.get(x, 0) + 1
+
+    for i in range(num_vowels, n):
+        x = random.choice(CONSONANTS)
+        hand[x] = hand.get(x, 0) + 1
+    while True:
+        litter = random.choice(VOWELS)
+        if litter in hand.keys() and hand[litter] == 1:
+            hand["*"] = hand.pop(litter)
+            return hand
+
+
+def update_hand(hand, word):
+    """
+    Does NOT assume that hand contains every letter in word at least as
+    many times as the letter appears in word. Letters in word that don't
+    appear in hand should be ignored. Letters that appear in word more times
+    than in hand should never result in a negative count; instead, set the
+    count in the returned hand to 0 (or remove the letter from the
+    dictionary, depending on how your code is structured).
+
+    Updates the hand: uses up the letters in the given word
+    and returns the new hand, without those letters in it.
+
+    Has no side effects: does not modify hand.
+
+    word: string
+    hand: dictionary (string -> int)
+    returns: dictionary (string -> int)
+    """
+    new_hand = hand.copy()
+    for litter in word.lower():
+        if litter in new_hand.keys() and new_hand[litter] > 0:
+            new_hand[litter] -= 1
+    return new_hand
+
+
+def is_valid_word(word, hand, word_list):
+    """
+    Returns True if word is in the word_list and is entirely
+    composed of letters in the hand. Otherwise, returns False.
+    Does not mutate hand or word_list.
+
+    word: string
+    hand: dictionary (string -> int)
+    word_list: list of lowercase strings
+    returns: boolean
+    """
+    word = word.lower()
+    if "*" in word:
+        for vowel in VOWELS:
+            if word.replace("*", vowel) in word_list:
+                return True
+        else:
+            return False
+    elif word in word_list:
+        for litter in word:
+            if (litter not in hand.keys()) or word.count(litter) > hand[litter]:
```

Comment:
You can remove one pair of brackets. Firstly Python process `not`, then `and`, then `or` logic operators.